pandas/core/dtypes/concat.py

```diff
@@ -21,11 +26,78 @@
 from pandas.core.construction import array, ensure_wrapped_if_datetimelike
 
 
+class NullArrayProxy:
+    """
+    Proxy object for an all-NA array.
+
+    Only stores the length of the array, and not the dtype. The dtype
+    will only be known when actually concatenating (after determining the
+    common dtype, for which this proxy is ignored).
+    Using this object avoids that the internals/concat.py needs to determine
+    the proper dtype and array type.
+    """
+
+    ndim = 1
+
+    def __init__(self, n: int):
+        self.n = n
+
+    @property
+    def shape(self):
+        return (self.n,)
+
+
+def _array_from_proxy(arr, dtype: DtypeObj, fill_value=lib.no_default):
+    """
+    Helper function to create the actual all-NA array from the NullArrayProxy object.
+
+    Parameters
+    ----------
+    arr : NullArrayProxy
+    dtype : the dtype for the resulting array
+    fill_value : scalar NA-like value
+        By default uses the ExtensionDtype's na_value or np.nan. For numpy
+        arrays, this can be overridden to be something else (eg None).
+
+    Returns
+    -------
+    np.ndarray or ExtensionArray
+    """
+    if is_extension_array_dtype(dtype):
+        return dtype.construct_array_type()._from_sequence(
+            [dtype.na_value] * arr.n, dtype=dtype
```

Comment:
We should maybe consider adding a method to the ExtensionArray interface to create an empty/all-NA array of the given dtype, to avoid this rather inefficient "construct from list of NA scalars".

Reply:
IIRC this can break for dtypes that cant hold NA.  the more robust version is 

```
cls = dtype.construct_array_type()
empty = cls._from_sequence([], dtype=dtype)
indexer = -np.ones(arr., dtype=np.intp)
return empty.take(indexer)
```

(in internals concat we use this in one place and when i tried to change it to what you have here something broke, dont remember what off the top of my head)

Reply:
Do we have an example dtype that cannot hold NAs?

Reply:
something like Sparse[int]

Reply:
Indeed:

```
In [5]: dtype = pd.SparseDtype("int")

In [6]: dtype
Out[6]: Sparse[int64, 0]

In [7]: dtype.construct_array_type()._from_sequence([dtype.na_value] * 5, dtype=dtype)
...
ValueError: Cannot convert non-finite values (NA or inf) to integer
```

Will use your suggested code for now

Reply:
I opened https://github.com/pandas-dev/pandas/issues/39776 for the idea of adding a method to the EA interface for creating an empty/all-NA array from a given dtype.